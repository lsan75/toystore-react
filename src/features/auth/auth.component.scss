.AuthComponent {
  background: #111;
  display: none;
  height: 25rem;
  position: absolute;
  right: 1rem;
  top: calc(100% + 1px);
  width: 18rem;
  &.opened {
    display: block;
  }
  z-index: 999;
}