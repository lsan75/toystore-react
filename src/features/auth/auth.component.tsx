import * as React from 'react';
import './auth.component.css';

interface Props {
  isOpened: boolean;
}
export function AuthComponent(props: Props) {
  const klass = props.isOpened ? 'opened' : '';

  return <div className={`AuthComponent ${klass}`} />;

}
