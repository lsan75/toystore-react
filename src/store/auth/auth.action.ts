import { Action } from 'redux';
import { AuthAction } from './auth.reducer';
import { UserState } from '../../features/auth/auth.component';
import { Dispatch } from 'react-redux';
import axios, { AxiosResponse } from 'axios';

export const AUTH = {
  SET_CONNECTED: 'AUTH_ET_CONNECTED',
  OPEN: 'AUTH_OPEN',
  SET_ERROR: 'AUTH_SET_ERROR'
};

export function authConnectAction(): Action {
  return {
    type: AUTH.SET_CONNECTED
  };
}

export function openAuthAction(): Action {
  return {
    type: AUTH.OPEN
  };
}

export function setError(payload: boolean): AuthAction {
  return {
    type: AUTH.SET_ERROR,
    payload
  };
}

export function submitAuthAction(sentUser: UserState) {
  return (dispatch: Dispatch<AuthAction>) => {

    let { user = '', pass = '' } = { ...sentUser };

    return axios.get(`/users?user=${user}&pass=${pass}`).then((result: AxiosResponse<UserState[]>) => {

      if (result.data.length) {
        dispatch(authConnectAction());
        dispatch(openAuthAction());
      } else {
        dispatch(setError(true));
      }

    }).catch(err => Promise.reject(err));

  };
}
