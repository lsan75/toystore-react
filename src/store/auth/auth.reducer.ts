import { Action } from 'redux';
import { AUTH } from './auth.action';

export interface AuthReducerState {
  isConnected: boolean;
  isOpened: boolean;
}

export const defaultAuthReducer = {
  isConnected: false,
  isOpened: false
};

export function authReducer(state: AuthReducerState = defaultAuthReducer, action: Action) {
  switch (action.type) {

    case AUTH.SET_CONNECTED:
      return { ...state, isConnected: true };

    case AUTH.OPEN:
      return { ...state, isOpened: !state.isOpened };

    default:
      return state;
  }
}