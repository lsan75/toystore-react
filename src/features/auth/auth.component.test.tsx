import * as React from 'react'
import { configure, shallow, mount, ShallowRendererProps, ShallowWrapper, ReactWrapper } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import { shallowToJson, mountToJson } from 'enzyme-to-json'

import { AuthComponent, Props, UserState } from './auth.component'

configure({ adapter: new Adapter() })

describe('AuthComponent', () => {

  let reactOutput: ReactWrapper<Props, UserState>

  const submitForm = jest.fn()

  beforeEach(() => {
    reactOutput = mount<Props, UserState>(

      <AuthComponent
        isOpened={true}
        isError={false}
        submit={submitForm}
      />

    )
  })

  it('renders without crashing', () => {
    const innerProps: Props = reactOutput.props()

    expect(mountToJson(reactOutput)).toMatchSnapshot()
    expect(innerProps.isOpened).toBe(true)
    expect(innerProps.isError).toBe(false)

    const wrapper = reactOutput.find('.AuthComponent')
    expect(wrapper.hasClass('opened')).toBe(true)
  })

  it('should submit the form', () => {
    const button = reactOutput.find('.AuthButton')

    const user = reactOutput.find('.AuthInput[type="email"]')
    const pass = reactOutput.find('.AuthInput[type="password"]')

    user.simulate('change', { target: { value: 'toto', id: 'user' }})
    pass.simulate('change', { target: { value: 'hophop', id: 'pass' }})
    pass.simulate('change', { target: { value: 'ignored', id: 'other' }})

    button.simulate('submit')

    expect(submitForm).toBeCalledWith({
      user: 'toto',
      pass: 'hophop'
    })
  })

  it('should be close', () => {
    reactOutput.setProps({ isOpened: false })

    const wrapper = reactOutput.find('.AuthComponent')
    expect(wrapper.hasClass('opened')).toBe(false)
  })
})
