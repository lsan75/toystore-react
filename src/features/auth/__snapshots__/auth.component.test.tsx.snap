// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthComponent renders without crashing 1`] = `
<AuthComponent
  isError={false}
  isOpened={true}
  submit={[MockFunction]}
>
  <div
    className="AuthComponent opened"
  >
    <form
      className="AuthForm"
      onSubmit={[Function]}
    >
      <h2
        className="AuthTitle"
      >
        Please Login
      </h2>
      <section
        className="AuthSection"
      >
        <label
          className="AuthLabel"
          htmlFor="user"
        >
          Your Email
        </label>
        <input
          className="AuthInput"
          id="user"
          onChange={[Function]}
          type="email"
          value=""
        />
      </section>
      <section
        className="AuthSection"
      >
        <label
          className="AuthLabel"
          htmlFor="pass"
        >
          Your Password
        </label>
        <input
          className="AuthInput"
          id="pass"
          onChange={[Function]}
          type="password"
          value=""
        />
      </section>
      <section
        className="AuthSectionButton"
      >
        <input
          className="AuthButton"
          type="submit"
          value="Submit"
        />
      </section>
    </form>
  </div>
</AuthComponent>
`;
